src/dataflow/src/render/join/delta_join.rs

```diff
@@ -199,93 +356,106 @@ where
+                                                local.enter_region(region),
+                                                stream_keys,
+                                                closure,
+                                            ),
+                                            Err(trace) => build_lookup(
+                                                update_stream,
+                                                trace.enter_region(region),
+                                                stream_keys,
+                                                closure,
+                                            ),
                                         }
                                     } else {
-                                        match arrangements_neu.get(&(&inputs[*other], &next_key[..])).unwrap() {
-                                            Ok(local) => build_lookup(update_stream, local.enter_region(region), prev_key, closure),
-                                            Err(trace) => build_lookup(update_stream, trace.enter_region(region), prev_key, closure),
+                                        match arrangements_neu
+                                            .get(&(&inputs[lookup_relation], &lookup_keys[..]))
+                                            .unwrap()
+                                        {
+                                            Ok(local) => build_lookup(
+                                                update_stream,
+                                                local.enter_region(region),
+                                                stream_keys,
+                                                closure,
+                                            ),
+                                            Err(trace) => build_lookup(
+                                                update_stream,
+                                                trace.enter_region(region),
+                                                stream_keys,
+                                                closure,
+                                            ),
                                         }
                                     };
                                     update_stream = oks;
                                     region_errs.push(errs);
-
-                                    bound_inputs.push(*other);
                                 }
 
                                 // We have completed the join building, but may have work remaining.
                                 // For example, we may have expressions not pushed down (e.g. literals)
                                 // and projections that could not be applied (e.g. column repetition).
-                                let closure = join_build_state.complete();
-                                if !closure.is_identity() {
+                                if !final_closure.is_identity() {
```

Comment:
you could lift this check into the delta path plan and have it store a `Option<JoinClosure>` (maybe also for the `initial_closure` too?)

Reply:
Could do! I have some other things to do here (fuse `final_closure` onto the last join) that I was going to hold back until after the "refactoring". Lemme see how this looks.